### PR TITLE
New routine_registration_linter

### PR DESCRIPTION
**Pull request by MichaelChirico**
Part of #884

### CODE REVIEW

File: tests/testthat/test-routine_registration_linter.R [hunk 1]
```diff
@@ -0,0 +1,16 @@
+test_that("routine_registration_linter skips allowed usages", {
```
Comment (IndrajeetPatil) [inline tests/testthat/test-routine_registration_linter.R:1]: Let's also use parametrized unit testing for allowed usages?

File: tests/testthat/test-routine_registration_linter.R [hunk 2]
```diff
@@ -0,0 +1,15 @@
+skip_if_not_installed("patrick")
+patrick::with_parameters_test_that(
+  "lints correctly",
```
Comment (IndrajeetPatil) [inline tests/testthat/test-routine_registration_linter.R:3]:
```suggestion
  "routine_registration_linter skips allowed usages and blocks disallowed usages correctly",
```

File: R/routine_registration_linter.R [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    xml <- source_expression$xml_parsed_content
+
+    bad_expr <- xml2::xml_find_all(xml, xpath)
+
+    xml_nodes_to_lints(
+      bad_expr,
+      source_expression = source_expression,
+      lint_message = "Register your native code routines with useDynLib.",
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/routine_registration_linter.R]: Good idea. Originally, our message points to some internal documentation, so now the message was a bit sparse. I think just mentioning `useDynLib` and `R_registerRoutines`, along with the full WRE link in `?routine_registration_linter`, should suffice

[the remark this answers is not in the document]
Comment (IndrajeetPatil) [reply R/routine_registration_linter.R]: SGTM!

**Review by IndrajeetPatil — APPROVED** [review]

**Pull request merged**